internal/context/workspace/domain/workflow/reader.go

```diff
@@ -0,0 +1,383 @@
+
+// GetWorkflowInputs ...
+func (r *WDLReader) GetWorkflowInputs(ctx context.Context, WorkflowFilePath string) ([]WorkflowParam, error) {
+	return r.getWorkflowParams(ctx, "java", "-jar", r.options.WomtoolPath, "inputs", WorkflowFilePath)
+}
+
+// GetWorkflowOutputs ...
+func (r *WDLReader) GetWorkflowOutputs(ctx context.Context, WorkflowFilePath string) ([]WorkflowParam, error) {
+	return r.getWorkflowParams(ctx, "java", "-jar", r.options.WomtoolPath, "outputs", WorkflowFilePath)
+}
+
+// GetWorkflowGraph ...
+func (r *WDLReader) GetWorkflowGraph(ctx context.Context, WorkflowFilePath string) (string, error) {
+	graph, err := exec.Exec(ctx, CommandExecuteTimeout, ".", "java", "-jar", r.options.WomtoolPath, "graph", WorkflowFilePath)
+	if err != nil {
+		return "", err
+	}
+
+	return string(graph), nil
+}
+
+// getWorkflowParams ...
+func (r *WDLReader) getWorkflowParams(ctx context.Context, name string, arg ...string) ([]WorkflowParam, error) {
+	params := make([]WorkflowParam, 0)
+	outputsResult, err := exec.Exec(ctx, CommandExecuteTimeout, ".", name, arg...)
+	if err != nil {
+		return params, err
+	}
+	var outputsMap map[string]string
+	if err := json.Unmarshal(outputsResult, &outputsMap); err != nil {
+		return params, err
+	}
+
+	for paramName, value := range outputsMap {
+		paramType, optional, defaultValue := r.parseWorkflowParamValue(value)
+		param := WorkflowParam{
+			Name:     paramName,
+			Type:     paramType,
+			Optional: optional,
+		}
+		if defaultValue != nil {
+			param.Default = *defaultValue
+		}
+		params = append(params, param)
+	}
+	// keep the return sort stable
+	sort.Slice(params, func(i, j int) bool {
+		return params[i].Name < params[j].Name
+	})
+	return params, nil
+}
+
+func (_ *WDLReader) parseWorkflowParamValue(value string) (paramType string, optional bool, defaultValue *string) {
+	splitByLeftBracket := strings.SplitN(value, "(", 2)
+	paramType = strings.TrimSpace(splitByLeftBracket[0])
+	if len(splitByLeftBracket) == 1 {
+		return paramType, false, nil
+	}
+
+	extraInfo := strings.TrimSuffix(splitByLeftBracket[1], ")")
+	splitByComma := strings.SplitN(extraInfo, ",", 2)
+	if strings.TrimSpace(splitByComma[0]) == "optional" {
+		optional = true
+	}
+	if len(splitByComma) == 1 {
+		return paramType, optional, nil
+	}
+
+	defaultInfo := strings.TrimSpace(splitByComma[1])
+	splitByEqual := strings.SplitN(defaultInfo, "=", 2)
+	if len(splitByEqual) != 2 || strings.ToLower(strings.TrimSpace(splitByEqual[0])) != "default" {
+		return paramType, optional, nil
+	}
+	rawDefaultValue := strings.TrimSpace(splitByEqual[1])
+	defaultValue = new(string)
+	if strings.HasPrefix(rawDefaultValue, `"`) && strings.HasSuffix(rawDefaultValue, `"`) { // String type
+		_ = json.Unmarshal([]byte(rawDefaultValue), defaultValue) // escape, never error
+	} else {
+		*defaultValue = rawDefaultValue
+	}
+	return paramType, optional, defaultValue
+}
+
+// NextflowReader reader for workflow written by Nextflow
+type NextflowReader struct {
```

Comment:
WDLReader / NextflowReader split by file would be better

Reply:
Separating WDLReader and NextflowReader into different files is done.  The names of the two files are wdl_reader.go and nextflow_readergo, respectively.